### PR TITLE
provide screen reader with inline suggestions

### PR DESCRIPTION
**Pull request by meganrogge**
cc @isidorn

### CODE REVIEW

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts [hunk 1]
```diff
@@ -156,7 +157,9 @@ export class GhostTextWidget extends Disposable {
 		this.partsWidget.setParts(ghostText.lineNumber, inlineTexts,
 			hiddenTextStartColumn !== undefined ? { column: hiddenTextStartColumn, length: textBufferLine.length + 1 - hiddenTextStartColumn } : undefined);
 		this.additionalLinesWidget.updateLines(ghostText.lineNumber, additionalLines, ghostText.additionalReservedLineCount);
-
+		const lineText = this.editor.getModel()!.getLineContent(ghostText.lineNumber);
+		console.log('should alert', ghostText.renderForScreenReader(lineText));
```
Comment (meganrogge) [inline src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:161]:
```suggestion
```

**Comment by jooyoungseo** [comment]
@meganrogge -- The best way for me to test this feature would be after it is released to insider because I don't know how to trigger inline suggestions like copilot in dev version. I usually test the inline suggestion with copilot extension, but as you know, I cannot use any extensions in dev version.

**Comment by meganrogge** [comment]
You can load the extension by downloading and loading the `.vsix` file, but I am able to test this so can proceed based on that

**Review by Tyriar — COMMENTED** [review]
You're probably best off getting a review from someone familiar with monaco for this one

File: src/vs/editor/common/standalone/standaloneEnums.ts [hunk 2]
```diff
@@ -179,140 +179,141 @@ export enum EditorOption {
...
-	wrappingIndent = 131,
-	wrappingStrategy = 132,
-	showDeprecated = 133,
-	inlayHints = 134,
-	editorClassName = 135,
-	pixelRatio = 136,
-	tabFocusMode = 137,
-	layoutInfo = 138,
-	wrappingInfo = 139
+	screenReaderAnnounceInlineSuggestion = 6,
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/editor/common/standalone/standaloneEnums.ts:182]: Henning had reviewed and approved these changes before, but will do for safety

**Comment by blaise80** [comment]
Pa fout janm voye message sou email ça anko ok

**Comment by meganrogge** [comment]
@hediet when I tried this, I didn't see an issue with audio cues/ debouncing. Could you pls suggest the repro steps for the issue you mentioned in the other PR and/or what you'd suggest?

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts [hunk 3]
```diff
@@ -157,6 +160,14 @@ export class GhostTextWidget extends Disposable {
...
 
+		this.audioCueService.playAudioCue(AudioCue.inlineSuggestion).then(() => {
+			if (this.editor.getOption(EditorOption.screenReaderAnnounceInlineSuggestion)) {
+				const lineText = this.editor.getModel()?.getLineContent(ghostText.lineNumber);
+				if (lineText) {
+					alert(ghostText.renderForScreenReader(lineText));
+				}
+			}
+		});
 		if (0 < 0) {
```
Comment (hediet) [inline src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:171]: Note that you are making the assumption that the `update` method is only called on proper changes. However, that is not the semantics of the `update` method, it just makes sure the state is up to date.

Comment (meganrogge) [reply src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:171]: Ok is there some other place you'd recommend I put this? I find this code a little confusing to understand

Comment (meganrogge) [reply src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:171]: figured it out

**Review by Tyriar — DISMISSED** [review]

**Review by mjbvz — APPROVED** [review]

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts [hunk 4]
```diff
@@ -157,6 +158,17 @@ export class GhostTextWidget extends Disposable {
 			hiddenTextStartColumn !== undefined ? { column: hiddenTextStartColumn, length: textBufferLine.length + 1 - hiddenTextStartColumn } : undefined);
 		this.additionalLinesWidget.updateLines(ghostText.lineNumber, additionalLines, ghostText.additionalReservedLineCount);
 
+		if (notifyUser) {
+			this.audioCueService.playAudioCue(AudioCue.inlineSuggestion).then(() => {
+				if (this.editor.getOption(EditorOption.screenReaderAnnounceInlineSuggestion)) {
+					const lineText = this.editor.getModel()?.getLineContent(ghostText.lineNumber);
+					if (lineText) {
+						alert(ghostText.renderForScreenReader(lineText));
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:166]: No, we specifically want `assertive` here - I've discussed this at length with @jooyoungseo

File: src/vs/workbench/contrib/audioCues/browser/audioCueLineFeatureContribution.ts [hunk 5]
```diff
@@ -258,37 +256,3 @@ class BreakpointLineFeature implements LineFeature {
 		);
 	}
 }
-
-class InlineCompletionLineFeature implements LineFeature {
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/audioCues/browser/audioCueLineFeatureContribution.ts:262]: @hediet can weight in here, but the issue was using this class caused the audio cue to play after the inline suggestion was announced - rendering it useless as an indicator of the type of announcement

[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/audioCues/browser/audioCueLineFeatureContribution.ts:262]: we can also await feedback from screen reader users

[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/audioCues/browser/audioCueLineFeatureContribution.ts:262]: but in order to get feedback, this has to be in insider's so they can use copilot / not have to install the extension into oss

**Comment by jooyoungseo** [comment]
@meganrogge -- Thanks! I have tested this morning and have found some issues, but hard to explain. I will take some time next week to file an issue. cut To cut to the chase, the suggestion live event is read aloud twice (tested with NVDA).

**Pull request merged**